src/basic/MAT_FUSED_MUL_ADD-Hip.cpp

```diff
@@ -0,0 +1,308 @@
+{
+  const Index_type run_reps = getRunReps();
+  const Index_type iend = getActualProblemSize();  
+  const Index_type N = m_N;
+  constexpr Index_type Ne = m_Ne;
+  constexpr Index_type NeNe = m_Ne * m_Ne;
+
+  dim3 gridDim (1, 1, 1);
+  dim3 blockDimBuiltin(Ne, 4, 1);
+
+  MAT_FUSED_MUL_ADD_DATA_SETUP;
+
+  MAT_FUSED_MUL_ADD_DATA_INIT;
+
+  if (vid == Base_HIP) {
+
+    MAT_FUSED_MUL_ADD_DATA_SETUP_HIP;
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+      hipLaunchKernelGGL((mat_fused_mul_add_builtin), dim3(gridDim), dim3(blockDimBuiltin), 0, 0, A, B, D, iend);
+      hipErrchk( hipGetLastError() );
+    }
+    stopTimer();
+
+    MAT_FUSED_MUL_ADD_DATA_TEARDOWN_HIP;
+  
+  } else {
+    getCout() << "\n  MAT_FUSED_MUL_ADD : Unknown Hip variant id = " << vid
+              << std::endl;
+  }
+}
+template < size_t block_size >
+void MAT_FUSED_MUL_ADD::runHipVariantImpl(VariantID vid)
+{
+  const Index_type run_reps = getRunReps();
+  const Index_type iend = getActualProblemSize();  
+  const Index_type N = m_N;
+  constexpr Index_type Ne = m_Ne;
+  constexpr Index_type NeNe = m_Ne * m_Ne;
+
+  constexpr Index_type block_x = gpu_block_size::sqrt(block_size);
+  constexpr Index_type block_y = gpu_block_size::sqrt(block_size);
+  dim3 blockDim(block_x, block_y);
+  dim3 gridDim(static_cast<size_t>(RAJA_DIVIDE_CEILING_INT(Ne, block_size)),
+               static_cast<size_t>(RAJA_DIVIDE_CEILING_INT(Ne, block_size)), 
+               static_cast<size_t>(1));
+  MAT_FUSED_MUL_ADD_DATA_SETUP;
+
+  MAT_FUSED_MUL_ADD_DATA_INIT;
+
+  if (vid == Base_HIP) {
+
+    MAT_FUSED_MUL_ADD_DATA_SETUP_HIP;
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+      hipLaunchKernelGGL((mat_fused_mul_add<block_size>), dim3(gridDim), dim3(blockDim), 0, 0, A, B, D, iend);
+      hipErrchk( hipGetLastError() );
+    }
+    stopTimer();
+
+    MAT_FUSED_MUL_ADD_DATA_TEARDOWN_HIP;
+  } else if (vid == Lambda_HIP) {
+
+    MAT_FUSED_MUL_ADD_DATA_SETUP_HIP;
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      auto mat_fused_lamda =
+        [=] __device__ (Index_type ii, Index_type row, Index_type col) {
+            MAT_FUSED_MUL_ADD_BODY;
+        };        
+      hipLaunchKernelGGL((mat_fused_lam<block_size, decltype(mat_fused_lamda)>),
+                         dim3(gridDim), dim3(blockDim), 0, 0,
+                         iend, mat_fused_lamda);      
+    hipErrchk( hipGetLastError() );
+    }
+    stopTimer();
+
+    MAT_FUSED_MUL_ADD_DATA_TEARDOWN_HIP;
+
+  } else if (vid == RAJA_HIP) {
+
+    MAT_FUSED_MUL_ADD_DATA_SETUP_HIP;
+
+    startTimer();
+    RAJA::RangeSegment ii_range(0, (N/(Ne*Ne)));
+    RAJA::RangeSegment row_range(0, Ne);
+    RAJA::RangeSegment col_range(0, Ne);
+
+    using EXEC_POL =
+      RAJA::KernelPolicy<
+        RAJA::statement::HipKernel<
+        RAJA::statement::For<2, RAJA::loop_exec,
```

Comment:
@MrBurmark Is there an easy RAJA way to make this loop_exec just go over these extents in parallel?

Reply:
@CRobeck You should be able to change it to hip_block_z_loop.  

Also, you should probably change the 'loop' policies in line 194 and 195 to 'direct' policies for performance; i.e., 'hip_block_y_loop' --> 'hip_block_y_direct'. Typically, we map a loop over tiles to block-direct policies and a loop withing a tile to a thread-direct policy.

Reply:
Here's an example from another kernel in the perf suite: https://github.com/LLNL/RAJAPerf/blob/develop/src/polybench/POLYBENCH_HEAT_3D-Hip.cpp#L179